Add --diff option teraslice-cli

sotojn opened the pull request:
This PR makes the following changes: 

- Adds `--diff` option to `teraslice cli view` that will compare the local `json` job file with what teraslice has on the state cluster
- Bumps `teraslice-cli` from `v2.1.0` to `v2.2.0`

Ref to issue #3694

busma13 commented on packages/teraslice-cli/src/helpers/jobs.ts:
```diff
@@ -669,6 +672,65 @@ export default class Jobs {
+        for (const file of files) {
+            const filePath = path.join(srcDir, file);
+            const jobConfig: JobConfigFile = JSON.parse(fs.readFileSync(filePath, { encoding: 'utf-8' }));
+            const formatedJobConfig = this.formatJobConfig(jobConfig);
+            localJobConfigs[formatedJobConfig.job_id as string] = formatedJobConfig;
```
```suggestion
            const formattedJobConfig = this.formatJobConfig(jobConfig);
            localJobConfigs[formatedJobConfig.job_id as string] = formattedJobConfig;
```

busma13 commented on packages/teraslice-cli/src/helpers/yargs-options.ts:
```diff
@@ -228,6 +228,11 @@ export default class Options {
             type: 'boolean'
         }),
+        diff: () => ({
+            describe: 'Shows diff on a job in a cluster and a job file',
```
Maybe `Shows diff between a job...`?

busma13 reviewed (approved):
Looks good

sotojn commented:
Here is an example on how this option looks. In the json file I change the amount of workers from 1 to 2 and run:

`earl tjm view test-job.json --diff`

Output:
```diff
  {
   "_context": "job",
   "_created": "2024-08-01T22:34:50.219Z",
-   "_updated": "2024-08-06T16:03:58.734Z",
+   "_updated": "2024-08-06T16:03:58.777Z",
    "assets": [
      "standard",
      "file"
    ],
    "job_id": "48e59745-ed75-4060-998e-002262abefc3",
    "lifecycle": "once",
    "name": "data-to-s3",
    "operations": [
      {
        "_op": "data_generator",
        "size": 100000
      },
      {
        "_op": "s3_exporter",
        "format": "ldjson",
        "path": "josephorigin"
      }
    ],
-   "workers": 1
+   "workers": 2
  }
```
```
--------
job: data-to-s3
id: 48e59745-ed75-4060-998e-002262abefc3
cluster: http://localhost:5678
--------
```
It does appear that updated seems slightly off all the time and maybe we shouldn't compare it?

busma13 commented:
> It does appear that updated seems slightly off all the time and maybe we shouldn't compare it?

I think it could be helpful to know that `_update` will be different between the two records.

sotojn commented:
> > It does appear that updated seems slightly off all the time and maybe we shouldn't compare it?
> 
> I think it could be helpful to know that `_update` will be different between the two records.

Should this be some sort of warning like, "_update field is subject to being slightly different from job file, to whats on the job record in the state cluster". Or should I just ignore this comparison entirely?

busma13 commented:
My feeling is that it's fine how it is now and a warning isn't necessary. But maybe it would be better to not compare `_update` if only the milliseconds are different.

ciorg commented:
If there is no difference could it out put something like - no difference between the configs instead of printing the whole config, not a big deal it's a pain to do so?   

Also, what color applies to what config? Green is the local file and red is the cluster config?  Maybe something to let people know what applies to what.

So far as the `update` difference, which for some reason is always different than the cluster value, at least on all the jobs I've looked at so far.   ( I'm not sure how the update is calculated, but it seems like that value should be the same for both?  Or maybe it shouldn't  even have millisecond precision?  But that is a different issue)   

It is interesting to know when the job was last updated, but if the diff is 100 ms it doesn't really tell me anything useful.  So maybe if there's a way to hide that diff unless it's at least a few min difference that would be better.

sotojn commented:
I've updated the behavior to include what is being compared to what. I also added logic around the "_updated" field to only display a diff if the time difference is greater than a minute. 

Results if "_updated field is less than 1 minute":
Command: `earl tjm view  data-to-s3.json  --diff`
```diff
  {
    "assets": [
      "standard",
      "file"
    ],
    "job_id": "48e59745-ed75-4060-998e-002262abefc3",
    "lifecycle": "once",
    "name": "data-to-s3",
    "operations": [
      {
        "_op": "data_generator",
        "size": 100000
      },
      {
        "_op": "s3_exporter",
        "format": "ldjson",
        "path": "josephorigin"
      }
    ],
-   "workers": 1   <--- state cluster value
+   "workers": 2   <--- local job file value
  }
```
```
--------
job: data-to-s3
id: 48e59745-ed75-4060-998e-002262abefc3
cluster: http://domain123
--------
```

Results if "_updated field is greater than 1 minute":
Command: `earl tjm view  data-to-s3.json  --diff`
```diff
  {
-   "_updated": "2024-08-06T16:03:58.734Z",   <--- state cluster value
+   "_updated": "2024-07-06T16:03:58.777Z",   <--- local job file value
    "assets": [
      "standard",
      "file"
    ],
    "job_id": "48e59745-ed75-4060-998e-002262abefc3",
    "lifecycle": "once",
    "name": "data-to-s3",
    "operations": [
      {
        "_op": "data_generator",
        "size": 100000
      },
      {
        "_op": "s3_exporter",
        "format": "ldjson",
        "path": "josephorigin"
      }
    ],
-   "workers": 1   <--- state cluster value
+   "workers": 2   <--- local job file value
  }
```
```
--------
job: data-to-s3
id: 48e59745-ed75-4060-998e-002262abefc3
cluster: http://domain123
--------
```

ciorg reviewed (approved):
LGTM

busma13 reviewed (approved):
Looks good.
